_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder.cpp

```diff
@@ -0,0 +1,1147 @@
+                    m_currHeaders.seqExtHdr = seqExtHdr;
+
+                    // Reset 'dependent' headers
+                    m_currHeaders.displayExtHdr.reset();
+                    m_currHeaders.qmatrix.reset();
+                    m_currHeaders.picHdr.reset();
+                    m_currHeaders.picExtHdr.reset();
+
+                    m_rawHeaders.Resize(unit.end - unit.begin);
+                    std::copy(unit.begin, unit.end, m_rawHeaders.GetPointer());
+                }
+                break;
+            case PICTURE_HEADER:
+                {
+                    size_t size = unit.end - unit.begin;
+                    auto picHdr = std::make_shared<MPEG2PictureHeader>();
+                    auto picExtHdr = std::make_shared<MPEG2PictureCodingExtension>();
+
+                    bitStream.GetPictureHeader(*picHdr.get());
+
+                    uint8_t* picExtBegin = RawHeaderIterator::FindStartCode(unit.begin + prefix_size + bitStream.BytesDecoded(), unit.end);
+                    if (!picExtBegin)
+                        return UMC::UMC_ERR_INVALID_STREAM;
+
+                    bitStream.Reset(picExtBegin + prefix_size, unit.end - picExtBegin);
+                    bitStream.SkipBits(8 + 4); // skip unit and extension types
+                    bitStream.GetPictureExtensionHeader(*picExtHdr.get());
+
+                    // Save headers
+                    m_currHeaders.picHdr = picHdr;
+                    m_currHeaders.picExtHdr = picExtHdr;
+                }
+                break;
+            case EXTENSION:
+            {
+                MPEG2_UNIT_TYPE_EXT type_ext = (MPEG2_UNIT_TYPE_EXT)bitStream.GetBits(4);
+                switch(type_ext)
+                {
+                case SEQUENCE_DISPLAY_EXTENSION:
+                    {
+                        auto displayExtHdr = std::make_shared<MPEG2SequenceDisplayExtension>();
+                        bitStream.GetSequenceDisplayExtension(*displayExtHdr.get());
+                        m_currHeaders.displayExtHdr = displayExtHdr;
+                    }
+                    break;
+                case QUANT_MATRIX_EXTENSION:
+                    {
+                        auto quantMatrix = std::make_shared<MPEG2QuantMatrix>();
+                        bitStream.GetQuantMatrix(*quantMatrix.get());
+                        m_currHeaders.qmatrix = quantMatrix;
+                    }
+                    break;
+                default:
+                    break;
+                }
+            }
+                break;
+            default:
+                break;
+            }
+        }
+        catch(const mpeg2_exception & ex)
+        {
+            return ex.GetStatus();
+        }
+        catch(...)
+        {
+            return UMC::UMC_ERR_INVALID_STREAM;
+        }
+
+        if (type == SEQUENCE_HEADER && m_firstMfxVideoParams.mfx.FrameInfo.Width)
+        {
+            const MPEG2SequenceHeader* currSeq = m_currHeaders.seqHdr.get();
```

Comment:
as I see, you don't need ```currSeq``` at all. smart pointers have everything to act like a pointer in this code